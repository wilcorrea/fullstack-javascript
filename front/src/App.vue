<template>
  <div id="app">
    <!-- where the magic happens -->
    <!-- @see front/src/router/index.ts:22 -->
    <router-view />
  </div>
</template>

<script lang="js">
import $store from 'src/store'
import Local from 'src/app/DataSource/Local'

/**
 */
export default {
  /**
   */
  name: 'App',
  /**
   */
  data: () => ({
    // state to map the version update confirmation
    confirmed: false
  }),
  /**
   */
  computed: {
    /**
     * Get the version from store
     * @see createStore
     * @returns {string}
     */
    version () {
      return $store.state.version
    }
  },
  /**
   */
  methods: {
    /**
     * This is a simple method to ask user to update the version
     * @param {string} message
     */
    notify (message) {
      if (this.confirmed) {
        return
      }
      this.confirmed = window.confirm(message)
      if (this.confirmed) {
        window.location.reload(true)
      }
    },
    /**
     * Receive the version fetched from server
     * @param {AxiosResponse} result
     */
    receiveVersion (result) {
      if (this.version && this.version !== result.data) {
        const message = `We have updates!` +
          ` The current version is ${this.version} and the new one is ${result.data}.` +
          ` Do you wanna reload the app now?`
        this.notify(message)
      }
      $store.commit('updateVersion', result.data)
    },
    /**
     * Fetch the current app version from server
     */
    fetchVersion () {
      Local
        .instance()
        .get('version')
        .then(this.receiveVersion)
    }
  },
  /**
   */
  created () {
    // wait 1 second to take a breath
    window.setTimeout(this.fetchVersion, 1000)
  }
}
</script>
