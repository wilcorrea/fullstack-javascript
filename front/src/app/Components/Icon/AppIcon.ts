import { CreateElement, RenderContext } from 'vue'

/**
 * @type {string}
 */
const block = 'app-icon'

/**
 * @component {AppIcon}
 */
export default {
  /**
   */
  functional: true,
  /**
   */
  name: 'AppIcon',
  /**
   * @param {CreateElement} h
   * @param {RenderContext} context
   */
  render (h: CreateElement, context: RenderContext) {
    const { props, listeners } = context
    const style = {}
    if (props.color) {
      // @ts-ignore
      style['color'] = props.color
    }
    if (listeners.click) {
      // @ts-ignore
      style['cursor'] = 'pointer'
    }
    const data = { class: `${block} material-icons`, style, on: { ...listeners } }
    let name = props.name
    if (!name) {
      name = 'warning'
      console.warn('[app-icon]', ' ', 'using a fallback icon, please provide a name')
    }

    return h('span', data, name)
  }
}
