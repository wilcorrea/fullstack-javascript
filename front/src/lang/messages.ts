import ptBr from 'src/lang/pt-br'
import { Record } from 'src/app/Agnostic/interfaces'

/**
 * The standard messages list
 * @type {Record}
 */
export const messages: Record = { 'pt-br': ptBr }

/**
 * Set the messages
 * Can be used to replace or add new messages
 * @param {string} alias
 * @param {Record} newLocale
 */
export const setMessages = (alias: string, newLocale: Record) => {
  messages[alias] = newLocale
}

/**
 * Get the messages
 * @param {string} alias
 * @returns {Record}
 */
export const getMessages = (alias: string): Record => {
  return messages[alias]
}

/**
 * Export getMessages as default
 * @param {string} locale
 * @returns {Record}
 */
export default getMessages
