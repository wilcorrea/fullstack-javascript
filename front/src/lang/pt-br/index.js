import app from './app'
import domains from 'src/domains/pt-br'
import pages from './pages'
import views from './views'

/**
 * put all messages together
 */
export default { app, domains, pages, views }
