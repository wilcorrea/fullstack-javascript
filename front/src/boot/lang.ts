import $lang from 'src/lang'

/**
 * @param {Vue} Vue
 * @returns {}
 */
export default (Vue: any) => {
  /**
   */
  Object.defineProperty(Vue.prototype, '$lang', {
    get () {
      return $lang
    }
  })
}
