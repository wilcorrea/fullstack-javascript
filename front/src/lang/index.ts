import messages from './messages'
import { get } from 'src/app/Util/general'
import { i18n } from 'src/settings'

/**
 * Start with a default locale
 * @type {string}
 */
let locale = i18n()

/**
 * Set the current locale
 * @param {string} newLocale
 */
export const setLocale = (newLocale: string) => {
  locale = newLocale
}

/**
 * Get the current locale
 * @returns {string}
 */
export const getLocale = (): string => {
  return locale
}

/**
 * Lang is a featured i18n engine
 * We can change the locale and add more messages in realtime
 * I do not like some things in vue-i18n and other libs that I've tested, so that's it
 * @param {string} path
 * @param {string} fallback
 * @returns {*}
 */
export default (path: string, fallback: string = ''): any => {
  return get(messages(locale), path, fallback)
}
