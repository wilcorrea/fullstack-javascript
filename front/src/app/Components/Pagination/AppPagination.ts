import { CreateElement, VNode } from 'vue'
import $lang from 'src/lang'

/**
 * @type {string}
 */
const block = 'app-pagination'

/**
 * @component {AppPagination}
 */
export default {
  /**
   */
  name: 'AppPagination',
  /**
   */
  props: {
    offset: {
      type: Number,
      required: true
    },
    limit: {
      type: Number,
      required: true
    }
  },
  data: () => ({
    page: 0
  }),
  /**
   */
  methods: {
    /**
     * @returns {number}
     */
    nextPage (): number {
      // @ts-ignore
      return this.limit * ++this.page
    },
    /**
     * @param {CreateElement} h
     * @returns {VNode}
     */
    renderButtonAppend (h: CreateElement): VNode {
      const data = {
        class: `${block}__append`,
        on: {
          click: () => {
            // @ts-ignore
            this.$emit('updateOffset', { offset: this.nextPage() })
          }
        }
      }
      const children = $lang('app.components.pagination.more')

      return h('button', data, children)
    },
    /**
     * @param {CreateElement} h
     * @returns {VNode}
     */
    renderButtonToTop (h: CreateElement): VNode {
      const data = {
        class: `${block}__to-top`,
        on: {
          click: () => {
            // @ts-ignore
            this.$emit('toTop')
          }
        }
      }
      const children = [h('app-icon', { props: { name: 'arrow_upward' } })]
      return h('button', data, children)
    }
  },
  /**
   * @param {CreateElement} h
   * @returns {VNode}
   */
  render (h: CreateElement): VNode {
    const data = { class: block }
    // @ts-ignore
    const children: Array<VNode> = [this.renderButtonAppend(h), this.renderButtonToTop(h)]
    return h('div', data, children)
  }
}
