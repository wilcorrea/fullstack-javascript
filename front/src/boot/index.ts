import lang from './lang'
import browse from './browse'
import util from './util'
import './polyfill'

/**
 * @type {{install: Function}}
 */
export default {
  install (Vue: any) {
    lang(Vue)
    browse(Vue)
    util(Vue)
  }
}
