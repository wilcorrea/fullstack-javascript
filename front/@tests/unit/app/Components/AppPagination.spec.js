import { shallowMount } from '@vue/test-utils'
import $lang from 'src/lang'
import AppPagination from 'src/app/Components/Pagination/AppPagination'
import AppIcon from 'src/app/Components/Icon/AppIcon'

describe('AppPagination', () => {
  const spy = jest.spyOn(console, 'error')
  shallowMount(AppPagination, { components: { AppIcon } })

  /**
   * @test
   */
  it('"offset" is required', () => {
    expect(spy).toBeCalledWith(expect.stringContaining('[Vue warn]: Missing required prop: "offset"'))
  })

  /**
   * @test
   */
  it('"limit" is required', () => {
    expect(spy).toBeCalledWith(expect.stringContaining('[Vue warn]: Missing required prop: "limit"'))
  })

  const offset = Math.round(Math.random() * 99)
  const limit = Math.round(Math.random() * 99)
  const wrapper = shallowMount(
    AppPagination,
    { components: { AppIcon }, propsData: { offset, limit } }
  )

  /**
   * @test
   */
  it('create markup with i18n correctly', () => {
    expect(wrapper.html()).toMatch('' +
      '<div class="app-pagination">' +
      '<button class="app-pagination__append">' +
      $lang('app.components.pagination.more') +
      '</button>' +
      '<button class="app-pagination__to-top">' +
      '<app-icon-stub name="arrow_upward"></app-icon-stub>' +
      '</button>' +
      '</div>')
  })

  /**
   * @test
   */
  it('test pagination management', () => {
    expect(wrapper.vm.offset).toBe(offset)
    expect(wrapper.vm.limit).toBe(limit)
    // noinspection JSUnresolvedFunction
    const nextOffset = wrapper.vm.nextPage()
    expect(wrapper.vm.page).toBe(1)
    expect(nextOffset).toEqual((wrapper.vm.page) * limit)
  })

  /**
   * @test
   */
  test('should render content correctly', () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })
})
