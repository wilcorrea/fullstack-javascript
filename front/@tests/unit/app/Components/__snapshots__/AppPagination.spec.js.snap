// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppPagination should render content correctly 1`] = `
<div
  class="app-pagination"
>
  <button
    class="app-pagination__append"
  >
    Mais resultados
  </button>
  <button
    class="app-pagination__to-top"
  >
    <app-icon-stub
      name="arrow_upward"
    />
  </button>
</div>
`;
